SYCL/ESIMD/api/functional/value.hpp

```diff
@@ -0,0 +1,183 @@
+  static_assert(CHAR_BIT == 8, "Unexpected byte size, input values may broke");
+  static_assert(size == 2, "Invalid number of bytes for half type");
+
+  const unsigned char in[size] = {lo, hi};
+  unsigned char index[size];
+  unsigned char out[size];
+
+  // We are using specific half value to initialize the bits required to differ
+  // the lowest and the highest byte
+  sycl::half indexHint = 2;
+  const unsigned char indexCoeff = 64;
+
+  memcpy(index, &indexHint, size);
+  index[0] /= indexCoeff;
+  index[1] /= indexCoeff;
+
+  // Ensure there is no overflow possible
+  assert(index[0] + index[1] == 1);
+
+  out[0] = in[index[0]];
+  out[1] = in[index[1]];
+  return esimd_test::bit_cast<sycl::half>(out);
+}
+
+} // namespace details
+
+// Utility class to retrieve specific values for tests depending on the data
+// type May be used to retrieve reference data or for generation of golden
+// values
+template <typename DataT> struct value {
+  static DataT inf() {
+    static_assert(
+        type_traits::is_sycl_floating_point_v<DataT>,
+        "Infinity is required only for the floating point data types.");
+
+    if constexpr (std::is_same_v<DataT, sycl::half>) {
+      return details::half_from_bytes(0b01111100u, 0b00000000u);
+    } else {
+      return std::numeric_limits<DataT>::infinity();
+    }
+  }
+
+  static DataT lowest() {
+    if constexpr (std::is_same_v<DataT, sycl::half>) {
+      return -max();
+    } else {
+      return std::numeric_limits<DataT>::lowest();
+    }
+  }
+
+  static DataT denorm_min() {
+    if constexpr (std::is_same_v<DataT, sycl::half>) {
+      return details::half_from_bytes(0b00000000u, 0b00000001u);
+    } else {
+      return std::numeric_limits<RefType>::denorm_min;
+    }
+  }
+
+  static DataT nan(unsigned char opcode = 42u) {
+    static_assert(type_traits::is_sycl_floating_point_v<DataT>,
+                  "NaN has meaning only for floating point data types.");
+    if constexpr (std::is_same_v<DataT, double>) {
+      return sycl::nan(static_cast<unsigned long>(opcode));
+    } else if constexpr (std::is_same_v<DataT, float>) {
+      return sycl::nan(static_cast<unsigned int>(opcode));
+    } else if constexpr (std::is_same_v<DataT, sycl::half>) {
+      return details::half_from_bytes(0b11111110u, 0b00000000u + opcode);
+    }
+  }
+
+  static DataT max() {
+    if constexpr (std::is_same_v<DataT, sycl::half>) {
+      return details::half_from_bytes(0b01111011u, 0b11111111u);
+    } else {
+      return std::numeric_limits<DataT>::max();
+    }
+  }
+};
+
+// Provides std::vector with the reference data according to the currently
+// tested data type and number of elements.
+template <typename DataT, int NumElems> std::vector<DataT> generate_ref_data() {
+  static_assert(
+      std::is_integral_v<DataT> || type_traits::is_sycl_floating_point_v<DataT>,
+      "Invalid data type provided to the generate_ref_data function.");
+
+  static const DataT min = value<DataT>::lowest();
```

Comment:
```suggestion
  // Create values with the strict type guarantee
  static const DataT min = value<DataT>::lowest();
```
Just as a suggestion - we may add a comment to avoid confusion about `min_plus_one` and `max_minus_one` values

Reply:
comment was added in this commit: 55ea733ea6bcac291953e9b4bf66a23c4ed7f275